backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceEventsHandler.java

```diff
@@ -0,0 +1,202 @@
+
+        long timeIntervalMillis = settings.getTimeIntervalMillis();
+
+        if (timeIntervalMillis != 0) {
+            final BacktraceEventsHandler handler = this;
+            handler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    handler.send();
+                    handler.postDelayed(this, timeIntervalMillis);
+                }
+            }, timeIntervalMillis);
+        }
+    }
+
+    public abstract void sendStartupEvent(String eventName);
+
+    public void send() {
+        sendEvents(events, 0);
+    }
+
+    private void sendEvents(ConcurrentLinkedDeque<Event> events, int attempts) {
+        if (events == null) {
+            return;
+        }
+        if (events.size() == 0) {
+            return;
+        }
+        try {
+            EventsPayload payload = getEventsPayload();
+            api.sendEventsPayload(payload);
+        } catch (Exception e) {
+            BacktraceLogger.e(LOG_TAG, "Could not create events payload for metrics submission");
+            return;
+        }
+    }
+
+    @Override
+    public void handleMessage(Message msg) {
+        final BacktraceHandlerInputEvents input = (BacktraceHandlerInputEvents) msg.obj;
+        EventsResult result;
+
+        if (input.eventsRequestHandler != null) {
+            BacktraceLogger.d(LOG_TAG, "Sending using custom request handler");
+            result = input.eventsRequestHandler.onRequest(input.payload);
+        } else {
+            BacktraceLogger.d(LOG_TAG, "Sending report using default request handler");
+            String json = BacktraceSerializeHelper.toJson(input.payload);
+            result = BacktraceReportSender.sendEvents(submissionUrl, json, input.payload, input.serverErrorEventListener);
+        }
+
+        if (input.eventsOnServerResponseEventListener != null) {
+            BacktraceLogger.d(LOG_TAG, "Processing result using custom event");
+            input.eventsOnServerResponseEventListener.onEvent(result);
+        }
+
+        int statusCode = result.getStatusCode();
+        if (statusCode == 200) {
+            onRequestCompleted();
+        }
+        else if (statusCode > 501 && statusCode != 505) {
+            numRetries++;
+            if (numRetries >= BacktraceMetrics.maxNumberOfAttempts || timeBetweenRetriesMillis == 0)
+            {
+                onMaximumAttemptsReached(input.payload.getEvents());
+                return;
+            }
+            final BacktraceEventsHandler handler = this;
+            handler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    EventsPayload payload = input.payload;
+                    payload.setDroppedEvents(numRetries);
+                    api.sendEventsPayload(payload);
+                }
+            }, calculateNextRetryTime(numRetries));
+        }
+    }
+
+    protected void onMaximumAttemptsReached(ConcurrentLinkedDeque<Event> events)
```

Comment:
Use generic type?

Reply:
This is a generic type. We are using a java collection ([equivalent to container](https://stackoverflow.com/a/9553091)) of our metrics base type `Event`

Reply:
I will mark this as resolved @vlussenburg